crates/xlinectl/src/command/watch.rs

```diff
@@ -67,23 +78,112 @@ pub(crate) async fn execute(client: &mut Client, matches: &ArgMatches) -> Result
+    let range_end = matches.get_one::<String>("range_end");
+    let request = build_request(matches)(key, range_end.map(String::as_str));
+
+    // extract the command provided by user
+    let command_to_execute: Vec<OsString> = matches
+        .get_many::<String>("exec_command")
+        .unwrap_or_default()
+        .map(OsString::from)
+        .collect();
+
+    let (_watcher, mut stream) = client.watch_client().watch(request).await?;
+    while let Some(resp) = stream
+        .message()
+        .await
+        .map_err(|e| XlineClientError::<XlineCommand>::WatchError(e.to_string()))?
+    {
+        resp.print();
+        if !command_to_execute.is_empty() {
+            execute_command_on_events(&command_to_execute, &resp)?;
         }
     }
 
     Ok(())
 }
 
+/// Execute user command for each event. We wanna support things like:
+///
+/// ```shell
+/// ./xlinectl watch foo -- sh -c "env | grep ETCD_WATCH_"
+/// ```
+///
+/// Expected output format:
+/// ```plain
+/// PUT
+/// foo
+/// bar
+/// XLINE_WATCH_REVISION=11
+/// XLINE_WATCH_KEY="foo"
+/// XLINE_WATCH_EVENT_TYPE="PUT"
+/// XLINE_WATCH_VALUE="bar"
+/// ```
+///
+fn execute_command_on_events(command_to_execute: &[OsString], resp: &WatchResponse) -> Result<()> {
+    let watch_revision = resp
+        .header
+        .as_ref()
+        .map(|header| header.revision)
+        .unwrap_or_default();
+
+    for event in &resp.events {
+        let kv = event.kv.as_ref().expect("expected key-value pair");
+        let watch_key = String::from_utf8(kv.key.clone()).expect("expected valid UTF-8");
+        let watch_value = String::from_utf8(kv.value.clone()).expect("expected valid UTF-8");
+        let event_type = match event.r#type {
+            0 => "PUT",
+            1 => "DELETE",
+            _ => "UNKNOWN",
+        };
+
+        let envs = HashMap::from([
+            ("XLINE_WATCH_REVISION", watch_revision.to_string()),
+            ("XLINE_WATCH_KEY", watch_key),
+            ("XLINE_WATCH_EVENT_TYPE", event_type.to_owned()),
+            ("XLINE_WATCH_VALUE", watch_value),
+        ]);
+
+        execute_inner(command_to_execute, envs)?;
+    }
+
+    Ok(())
+}
+
+/// Actual executor responsible for the user command,
+/// command format: [exec-command arg1 arg2 ...]
+#[allow(clippy::indexing_slicing)]
+fn execute_inner(command: &[OsString], envs: HashMap<&str, String>) -> Result<()> {
+    let mut cmd = StdCommand::new(&command[0]);
+
+    // adding environment variables
+    for (key, value) in envs {
+        let _ = cmd.env(key, value);
+    }
+
+    // collecting the args
+    if command.len() > 1 {
+        let _ = cmd.args(&command[1..]);
+    }
+    // executing the command
+    let output = cmd.output()?;
+    if !output.status.success() {
+        eprintln!("This command is not supported yet!");
+    }
+    println!("{}", String::from_utf8_lossy(&output.stdout));
+    eprintln!("{}", String::from_utf8_lossy(&output.stderr));
```

Comment:
Why do you print the output message twice?

Reply:
My original intention is like this : one for normal case (stdout), and one for error case (stderr). Now I have changed it into a better one.